content/engineering/guest-post/how-to-implement-jwt-authentication-in-deno/index.md

```diff
@@ -0,0 +1,652 @@
+│   │ 
+│   └───routes
+│   │
+│   └───schema
+│   │ 
+│   └───utils
+└─**app.ts**
+
+## Create Oak Server
+
+Next is to create your Oak server. [Oak](https://oakserver.github.io/oak/) is a Deno middleware system that provides a router middleware for HTTP servers.
+
+Go to `app.ts`, import Application from the Deno Oak URL, create an instance of the application, define your port, and call a middleware, as follows.
+
+`app.ts`
+
+```js
+import { Application } from "https://deno.land/x/oak/mod.ts";
+
+const app = new Application();
+const PORT = 8080;
+
+   app.use((ctx, next) => {
+       ctx.response.body = 'Welcome';
+        next();
+       });
+
+console.log(`Application is listening on port: ${PORT}`);
+
+await app.listen({port:PORT});
+```
+
+Let's quickly run your server: to run the server, use `deno run --allow-net app.ts` 
+
+Deno will always request permission to use your network. `--allow-net` gives Deno permission to all network calls.
+
+You should see `Application is listening on port: 8080`
+
+Now, in your `routes` folder, create a file called `allRoutes.ts`, and set up your router by importing `Router` from the oak url, then create an instance of the router and export the default router.
+
+`routes.allRoutes.ts`
+
+```js
+import { Router } from "https://deno.land/x/oak/mod.ts";
+
+const router = new Router();
+
+export default router;
+```
+
+Moving on, you need to modify `app.ts` file. So, go to `app.ts` and import the router. Let the app use `router.routes()` and `router.allowedMethods()` methods, then remove the middleware so it doesn't overshadow the imported routes. The `allowedMethods()` tells Deno to include all routes by your router.
+
+`app.ts`
+
+```js
+import { Application } from "https://deno.land/x/oak/mod.ts";
+import router from "./src/routes/allRoutes.ts";
+
+const app = new Application();
+const PORT = 8080;
+
+app.use(router.routes());
+app.use(router.allowedMethods());
+
+console.log(`Application is listening on port: ${PORT}`);
+
+await app.listen({port:PORT});
+```
+
+You can run your app again with the same command: `deno run --allow-net app.ts`
+
+## Create User Data
+
+Now, you need to create user data. But before that, let's connect your application to MongoDB database.
+
+- Go to your `database` folder
+- Create a file called `connectBD.ts`
+- Import MongoClient from the Deno MongoDB URL
+- Create an instance of MongoClient
+- Connect your database
+- Export default
+
+`database.connectDB.ts`
+
+```js
+import { MongoClient } from "https://deno.land/x/mongo@v0.30.0/mod.ts";
+
+  const client = new MongoClient();
+
+  await client.connect({
+    db: 'deno_auth',
+    tls: true,
+    servers: [
+      {
+        host: 'your-mongo-db-host',
+        port: 27017,
+      },
+    ],
+    credential: {
+      username: 'dbUserName',
+      password: 'dbPassword',
+      mechanism: 'SCRAM-SHA-1',
+      db: "admin",
+    },
+ })
+
+ console.log("Database connected!");
+ const db = client.database("deno_auth");
```

Comment:
@icode247  Unable to connect to DB. Facing following error:
Uncaught (in promise) Error: MongoError: "Connection failed: MongoError: {\"ok\":0,\"errmsg\":\"Invalid database name: ''\",\"code\":73,\"codeName\":\"InvalidNamespace\"}"

But it works when I updated this code by following code::
```
import { MongoClient } from "https://deno.land/x/mongo@v0.30.0/mod.ts";

const dbString = `DB_STRING`;
const client = new MongoClient();
await client.connect(dbString);
const db = client.database("deno_auth");
export default db;
```